miniupnpd/upnpstun.c

```diff
@@ -0,0 +1,464 @@
+			break;
+
+	}
+
+	/* Remove unblock for local ports */
+	for (i = 0; i < 4; ++i)
+		delete_filter_rule(if_name, local_ports[i], IPPROTO_UDP);
+
+	close(fds[0]);
+	close(fds[1]);
+	close(fds[2]);
+	close(fds[3]);
+
+	/* Parse received STUN messages */
+	have_ext_addr = 0;
+	for (i = 0; i < 4; ++i) {
+		if (parse_stun_response(responses[i], responses_lens[i], &mapped_addrs[i]) == 0)
+			have_mapped_addrs[i] = 1;
+		else
+			have_mapped_addrs[i] = 0;
+		if (!have_ext_addr && have_mapped_addrs[i]) {
+			memcpy(ext_addr, &mapped_addrs[i].sin_addr, sizeof(*ext_addr));
+			have_ext_addr = 1;
+		}
+	}
+
+	/* We have no external address */
+	if (!have_ext_addr) {
+		errno = ENXIO;
+		return -1;
+	}
+
+	for (i = 0; i < 4; ++i) {
+		if (!have_mapped_addrs[i]) {
+			/* We have not received all four responses, therefore NAT or firewall is doing some filtering */
+			*restrictive_nat = 1;
+			return 0;
+		}
+	}
+
+	if (memcmp(&remote_addr, &peer_addrs[0], sizeof(peer_addrs[0])) != 0) {
+		/* We received STUN response from different address even we did not asked for it, so some strange NAT is active */
+		*restrictive_nat = 1;
+		return 0;
+	}
+
+	for (i = 0; i < 4; ++i) {
+		if (ntohs(mapped_addrs[i].sin_port) != local_ports[i] || memcmp(&mapped_addrs[i].sin_addr, ext_addr, sizeof(*ext_addr)) != 0) {
+			/* External IP address or port was changed, therefore symmetric NAT is active */
+			*restrictive_nat = 1;
+			return 0;
+		}
+	}
+
+	/* Otherwise we are either directly connected or behind unrestricted NAT 1:1 */
+	/* There is no filtering, so port forwarding would work fine */
+	*restrictive_nat = 0;
+	return 0;
+}
+
+#ifdef TEST_LINUX_DEBUG_APP
+
+/* This linux test application for debugging purposes can be compiled as: */
+/* gcc upnpstun.c -o upnpstun -g3 -W -Wall -O2 -DTEST_LINUX_DEBUG_APP */
+
+#include <arpa/inet.h>
+#include <time.h>
+
+static int add_filter_rule2(const char *ifname, const char *rhost, const char *iaddr, unsigned short eport, unsigned short iport, int proto, const char *desc)
+{
+	char buffer[100];
+	ifname = ifname;
+	rhost = rhost;
+	iaddr = iaddr;
+	iport = iport;
+	desc = desc;
+	snprintf(buffer, sizeof(buffer), "/sbin/iptables -t filter -I INPUT -p %d --dport %hu -j ACCEPT", proto, eport);
+	printf("Executing: %s\n", buffer);
+	return system(buffer);
+}
+
+static int delete_filter_rule(const char * ifname, unsigned short port, int proto)
+{
+	char buffer[100];
+	ifname = ifname;
+	snprintf(buffer, sizeof(buffer), "/sbin/iptables -t filter -D INPUT -p %d --dport %hu -j ACCEPT", proto, port);
+	printf("Executing: %s\n", buffer);
+	return system(buffer);
+}
+
+int main(int argc, char *argv[])
+{
+	struct in_addr ext_addr;
+	int restrictive_nat;
+	int ret;
+	char str[INET_ADDRSTRLEN];
+
+	if (argc != 3 && argc != 2) {
+		printf("Usage: %s stun_host [stun_port]\n", argv[0]);
+		return 1;
+	}
+
+	if (argc == 2)
+		argv[2] = "0";
+
+	srandom(time(NULL) * getpid());
```

Comment:
srandom() is called in miniupnpd.c somewhere.

Reply:
oops sorry it is in the main() there :)

Reply:
It is undef TEST_LINUX_DEBUG_APP which produce testing stun client application. Useful for development and debugging.